62606 Return fully qualified path from FileSystemEventArgs

slask opened the pull request:
The FullPath for FileSystemEventArgs and related OldFullPath for RenamedEventArgs now return fully qualified paths as per the documentation and the name of these properties and according to the discussions in the issue below.

Fix #62606

slask committed:
FullPath property now returns fully qualified path

Gets the fully qualified path after combining the directory and name part
as per the thread discussion. Also added and updated tests for the new change.

Fix #62606

slask committed:
OldFullPath returns fully qualified path

To be consistent with new behavior of FullPath from FileSystemEventArgs
same change was applied for RenamedEventArgs including updated tests.

Fix #62606

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63051) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63051) before we can accept your contribution.<br/><br/>:x: slask [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63051)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=63051) it.</sub>

slask commented:
Hi @Jozkee . Based on the feedback from the failed tests I believe the fix is good but the tests are too Windows OS oriented. They were the same before my changes but back then it was not an issue since the FullPath was not actually returning a truly fully qualified path but mostly a simple concatenation of directory and file name.

Two options that I see now are:
- remove the windows specific test cases (_InlineData_ attribute instances)
- extract the windows specific test cases and mark them somehow to be run on Windows only. I might need some pointers on how to do that. Is it the _ActiveIssue_ attribute?

Please let me know your thoughts. Thanks!

slask committed:
Marked some test cases as Windows OS specific

In the context of using an actual full path
which can vary on Unix vs Windows

Fix #62606

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,71 @@ namespace System.IO.Tests
-            {
-                directory += Path.DirectorySeparatorChar;
-            }
+            directory = AppendDirectorySeparator(directory);
```
For `C:` we shouldn't append a separator to `directory` if there is not one. Such path is ["A relative path from the current directory of the C: drive."](https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats).

We should rather add another parameter `expectedFullPath` and assert that against `args.FullPath`. Thoughts?

slask replied:
Agree. Added in next commit.

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemEventArgs.cs:
```diff
@@ -19,7 +19,7 @@ public FileSystemEventArgs(WatcherChangeTypes changeType, string directory, stri
             _changeType = changeType;
             _name = name;
-            _fullPath = Combine(directory, name);
+            _fullPath = Path.GetFullPath(Combine(directory, name));
```
`Combine("", "")` will return `\`, which I think is a bug since a single separator refers to the root folder.
We should start using `Path.Join` instead to avoid the code duplication.

```suggestion
            _fullPath = Path.GetFullPath(Path.Join(directory, name));
```

slask replied:
Done.

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/src/System/IO/RenamedEventArgs.cs:
```diff
@@ -18,7 +18,7 @@ public RenamedEventArgs(WatcherChangeTypes changeType, string directory, string?
         {
             _oldName = oldName;
-            _oldFullPath = Combine(directory, oldName);
+            _oldFullPath = Path.GetFullPath(Combine(directory, oldName));
```
```suggestion
            _oldFullPath = Path.GetFullPath(Path.Join(directory, oldName));
```

slask replied:
Done.

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,71 @@ namespace System.IO.Tests
+        [Theory]
+        [InlineData((WatcherChangeTypes)0, "", "")]
+        [InlineData((WatcherChangeTypes)0, "", null)]
+        public static void FileSystemEventArgs_ctor_When_EmptyDirectory_Then_FullPathReturnsTheCurrentVolume(WatcherChangeTypes changeType, string directory, string name)
+        {
+            FileSystemEventArgs args = new FileSystemEventArgs(changeType, directory, name);
+
+            directory = AppendDirectorySeparator(directory);
+
+            Assert.Equal(changeType, args.ChangeType);
+            Assert.Equal(Directory.GetDirectoryRoot(directory + name), args.FullPath);
+            Assert.Equal(name, args.Name);
```
Based on what I said about `Combine("", "") == '\'`, I think that for these cases `args.FullPath` should be `string.Empty`.

slask replied:
Done.

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.RenamedEventArgs.cs:
```diff
@@ -16,26 +16,80 @@ public static void RenamedEventArgs_ctor(WatcherChangeTypes changeType, string d
         {
```
My comments in FileSystemEventArgs apply for RenamedEventArgs as well.

slask replied:
Done.

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemEventArgs.cs:
```diff
@@ -19,7 +19,7 @@ public FileSystemEventArgs(WatcherChangeTypes changeType, string directory, stri
             _changeType = changeType;
             _name = name;
-            _fullPath = Combine(directory, name);
+            _fullPath = Path.GetFullPath(Combine(directory, name));
```
Using `Path.GetFullPath` here will add an extra allocation every time `FileSystemWatcher` reports an event. We should consider caching the full path in FSW and use that to create a `FileSystemEventArgs`.

slask replied:
Hi @Jozkee. Hope you had a nice holiday!

FSW will notify every time the file changes so the same allocation will happen again but in FSW this time if I am not 
mistaken. Unless you are referring to getting the fully qualified path of the **directory** (<code>_directory</code>private field) used to initialize FSW in the constructor.

If that is the case, then some of the test cases for <code>FileSystemEventrArgs</code> will be rendered obsolete because the constructor's _directory_ parameter will always be a fully qualified path.

This means also that we are pushing this responsibility in FSW instead of the FSEventArgs but I guess that is Ok since from what I saw <code>FileSystemEventArgs</code> is not used **standalone** anywhere else.

Jozkee replied:
> Unless you are referring to getting the fully qualified path of the directory (_directoryprivate field) used to initialize FSW in the constructor.

Yes I was thinking on something more like this. 

As it is now, it is calling Path.GetFullPath for each event reported by the same instance of FileSystemWatcher. We could cache the result of GetFullPath in `_directory` and pass that to `FileSystemEventArgs` and `RenamedEventArgs` through new internal constructors (that can skip path normalization).

Although before we proceed, there is a caveat with caching the full path of a FileSystemWatcher; currently `_directory ` holds a relative path (relative to the CWD), if the CWD changes, I think that the watched folder will change as well. I'm not sure if that's an scenario that we should continue to support or not. If we decide to proceed with the caching, that scenario will be broken.

@adamsitnik @carlossanlop @ericstj @danmoseley thoughts?

slask replied:
@Jozkee I was thinking to keep the <code>_directory</code> as is and have another private filed <code>_directoryFullPath</code> that is written together with _directory (there are about 2-3 places). Then use the new full path dir only when instantiating the event args classes. Without interfering with how FSW works right now.

Jozkee replied:
That would also reflect the behavior change that I described above. e.g: FSW passes `_directory` to CreateFile
https://github.com/dotnet/runtime/blob/312c66f1fc2f749f56612999cb1adab9ca7fde59/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs#L31-L33

If we add _directoryFullPath there is a chance that `Path.GetFullPath(_directory) != _directoryFullPath` and hence `FileSystemEventsArgs.FullPath` would be incorrect.

Based on this discussion I suggest to not cache the full path and make FileSystemEventsArgs call GetFullPath always. What do you guys think?

slask replied:
Resolving this comment as per: https://github.com/dotnet/runtime/pull/63051#discussion_r783507896

Jozkee commented:
@slask thanks for your PR.

I just saw this:

> Two options that I see now are:
> 
> remove the windows specific test cases (InlineData attribute instances)
> extract the windows specific test cases and mark them somehow to be run on Windows only. I might need some pointers on how to do that. Is it the [ActiveIssue(...)] or [PlatformSpecific(TestPlatforms.Windows)] attribute?

Second option is the way to go, which based on what I saw you already did correctly.

slask committed:
Using Path.Join instead of the custom Combine

Adjusted unit tests to match the new logic.

Fix #62606

slask committed:
Account for corner case when fullPath matches root

The previous buggy implementation of RenamedEventArgs
produced a convenient side-effect for PhysicalFilesWatcher.

Without the trailing slash adde by RenamedEventArgs the root
was longer than the fullPath by a trailing slash.

Fix #62606

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,61 @@ namespace System.IO.Tests
-        [InlineData(WatcherChangeTypes.Changed, "C:", "foo.txt")]
-        [InlineData(WatcherChangeTypes.All, "C:", "foo.txt")]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        [InlineData(WatcherChangeTypes.Changed, "C:", "foo.txt", "C:\\foo.txt")]
```
This case is not right, `expectedFullPath` should be `{yourCurrentWorkingDirectory}\foo.txt`.

The test is passing because you are calling `Path.Join(directory, name)` before `Path.GetFullPath(directory)` and the former method doesn't treat `C:` as "your CWD in drive C", it is just concatenating both paths with a `\` in the middle.

The fix should be very simple, flip the order that the methods are called. I envision that this will now break the test cases where you pass string.Empty as directory, that is ok, i.e: it is OK that Path.GetFullPath throws for an empty path.

slask replied:
Fixed in the latest commit (f28bf24)

slask replied:
Hi @Jozkee. Are you happy with the changes? Can this comment be resolved?

I haven't seen any answer in a week to the above thread with moving on without caching. Should we wait longer?

Jozkee replied:
We can move on, thanks for fixing it.
Sorry for the delay, I have been heads down addressing other work items, I'm still interested in moving forward with your PR if you are as well.

slask replied:
Yes, of course. I will implement your feedback soon.

slask committed:
Handle relative paths from CWD in given drive

Fix #62606

Jozkee commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:
```diff
@@ -340,6 +340,15 @@ private void OnFileSystemEntryChange(string fullPath)
         {
             try
             {
+                // this can happen when the fullPath matches the root directory path
+                // but the root path is always created with a trailing slash
+                // this was masked by a side effect (bug) in RenamedEventArgs that was appending the trailing slash to fullPath value
+                // that behavior was changed with the fix for https://github.com/dotnet/runtime/issues/62606
+                if (fullPath.Length < _root.Length)
+                {
+                    return;
+                }
```
Can you please ensure `Renamed/FileSystemEventArgs.FullPath` returns with a trailing separator if the `[old]Name` argument is null or empty. 
I'm sure we want to keep that specific part of the old behavior.

Let me know if you need help on doing that and I can give you some pointers.

slask replied:
Fixed in commit aaf4b45

slask replied:
@Jozkee this was the final item on the review list. Tests pass and the previous failures did not seem related and could have been transient issues. Thanks!

slask committed:
Clearly separated unix and windows test cases

Implemented most of the code review suggestions.

Fix #62606

slask committed:
Merge branch 'main' of https://github.com/dotnet/runtime into 62606-return-fully-qualified-path-FileSystemEventArgs

slask commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,86 @@ namespace System.IO.Tests
+        [Theory]
+        [PlatformSpecific(TestPlatforms.Windows)]
         [InlineData(WatcherChangeTypes.Changed, "C:", "foo.txt")]
         [InlineData(WatcherChangeTypes.All, "C:", "foo.txt")]
```
Fixed in commit a025928

slask commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,86 @@ namespace System.IO.Tests
+        public static void FileSystemEventArgs_ctor_Invalid_NullDirectory()
+        {
+            Assert.Throws<ArgumentNullException>(() => new FileSystemEventArgs((WatcherChangeTypes)0, null, "foo.txt"));
+        }
```
Fixed in commit a025928

slask commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,86 @@ namespace System.IO.Tests
+
+        [Theory]
+        [InlineData(WatcherChangeTypes.Changed, "bar", "foo.txt")]
+        [InlineData(WatcherChangeTypes.Changed, "bar\\baz", "foo.txt")]
```
Fixed in commit a025928

slask commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,86 @@ namespace System.IO.Tests
+        [InlineData(WatcherChangeTypes.Changed, "E:\\bar", "foo.txt", "E:\\bar\\foo.txt")]
+        [InlineData(WatcherChangeTypes.All, "D:\\", "foo.txt", "D:\\foo.txt")]
+        [InlineData(WatcherChangeTypes.All, "E:\\bar", "foo.txt", "E:\\bar\\foo.txt")]
+        public static void FileSystemEventArgs_ctor_AbsolutePaths(WatcherChangeTypes changeType, string directory, string name, string expectedFullPath)
```
Fixed in commit a025928

slask committed:
Use EnsureTrailingSeparator instead of custom code

Also made unit tests more focused by
removing needless asserts from the tests

Fix #62606

slask commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,28 +8,86 @@ namespace System.IO.Tests
+
+        #region Test Helpers
+
+        private static string AppendDirectorySeparator(string directory)
```
Fixed in commit 5ff61d5

slask committed:
Keeping the old behavior with trailing separator

Reverted the changes to PhysicalFilesWatcher
and added back the trailing separator behavior
as per code review suggestion

Fix #62606

slask commented:
Linking doc issue created for breaking change: https://github.com/dotnet/docs/issues/27916

slask commented:
@Jozkee I created the breaking change doc issue and cross-linked them. I also emailed dotnetbcn@microsoft.com with the docs issue.

I don't think I have permission to remove the tag from this PR.

slask commented:
Hi @Jozkee, is there something else I need to do to finalize this PR?

slask commented:
Hi @Jozkee. Friendly reminder to move this PR forward.

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,86 +8,93 @@ namespace System.IO.Tests
-        [InlineData(WatcherChangeTypes.All, "D:\\", "foo.txt", "D:\\foo.txt")]
-        [InlineData(WatcherChangeTypes.All, "E:\\bar", "foo.txt", "E:\\bar\\foo.txt")]
-        public static void FileSystemEventArgs_ctor_AbsolutePaths(WatcherChangeTypes changeType, string directory, string name, string expectedFullPath)
+        [InlineData(WatcherChangeTypes.Changed, "C:\\bar", "foo.txt")]
+        [InlineData(WatcherChangeTypes.All, "C:\\bar", "foo.txt")]
+        [InlineData((WatcherChangeTypes)0, "C:\\bar", "")]
+        [InlineData((WatcherChangeTypes)0, "C:\\bar", null)]
```
Unix will treat C:\ as a relative path; better use something that is relative in both OSes. 
```suggestion
        [InlineData(WatcherChangeTypes.Changed, "bar", "foo.txt")]
        [InlineData(WatcherChangeTypes.All, "bar", "foo.txt")]
        [InlineData((WatcherChangeTypes)0, "bar", "")]
        [InlineData((WatcherChangeTypes)0, "bar", null)]
```

slask replied:
fixed

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,86 +8,93 @@ namespace System.IO.Tests
-            Assert.Throws<ArgumentException>(() => new FileSystemEventArgs((WatcherChangeTypes)0, "", "foo.txt"));
+            FileSystemEventArgs args = new FileSystemEventArgs(WatcherChangeTypes.All, directory, name);
+
+            directory = PathInternal.EnsureTrailingSeparator(directory);
+
+            Assert.Equal(PathInternal.EnsureTrailingSeparator(Directory.GetCurrentDirectory()) + directory + name, args.FullPath);
```
Path.Combine will ensure the trailing separator.

```suggestion

            string expectedPath = Path.Combine(Directory.GetCurrentDirectory(), directory, name);
            Assert.Equal(expectedPath, args.FullPath);
```

slask replied:
fixed

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,86 +8,93 @@ namespace System.IO.Tests
-            Assert.Throws<ArgumentNullException>(() => new FileSystemEventArgs((WatcherChangeTypes)0, null, "foo.txt"));
+            FileSystemEventArgs args = new FileSystemEventArgs(WatcherChangeTypes.All, directory, name);
+
+            directory = PathInternal.EnsureTrailingSeparator(directory);
+
+            Assert.Equal(PathInternal.EnsureTrailingSeparator(Directory.GetCurrentDirectory()) + directory + name, args.FullPath);
```
```suggestion
            string expectedPath = Path.Combine(Directory.GetCurrentDirectory(), directory, name);
            Assert.Equal(expectedPath, args.FullPath);
```

slask replied:
fixed

Jozkee commented on src/libraries/System.IO.FileSystem.Watcher/tests/Args.FileSystemEventArgs.cs:
```diff
@@ -8,86 +8,93 @@ namespace System.IO.Tests
-            Assert.Equal(changeType, args.ChangeType);
-            Assert.Equal(AppendDirectorySeparator(Directory.GetCurrentDirectory()) + directory, args.FullPath);
-            Assert.Equal(name, args.Name);
+            Assert.Equal(PathInternal.EnsureTrailingSeparator(Directory.GetCurrentDirectory()) + name, args.FullPath);
```
```suggestion
            string expectedPath = Path.Combine(Directory.GetCurrentDirectory(), name);
            Assert.Equal(expectedPath, args.FullPath);
```

slask replied:
Fixed

jozkee reviewed (approved):
Just minor nits in test code; otherwise, looks good to me, thanks.

slask committed:
Used Path.Combien vs PathInternalEnsureSeparator

As per code review suggestions.

Fix #62606

slask committed:
Added blank line for readability

Pipeline retrigger. Some pipelines timed out, most probably transient issues.

Fix #62606

Jozkee commented:
@slask, thanks for your contribution!

danmoseley commented:
#68883
